packages/protocol/contracts/common/AddressResolver.sol

```diff
@@ -69,6 +69,16 @@ abstract contract AddressResolver {
         return address(_addressManager);
     }
 
+    function keyForName(
+        uint256 chainId,
+        string memory name
+    ) public pure returns (string memory key) {
+        key = string.concat(Strings.toString(chainId), ".", name);
+        // TODO: the next line is cheaper in gas but will break
+        //       many Hardhat tests.
+        // key = string(bytes.concat(bytes32(chainId), bytes(name)));
```

Comment:
Hey @dantaik !

The reason why the hardhat tests are failing (and foundry not) is because the `bytes32(uint256)` kind of conversion does convert the integer into a seemingly empty bytes variable - which only on the surface looks empty.
So basically here, what you get back is the `EMPTY_BYTES + bytes(name)` , within foundry it is passing because the expects also written in Solidity, but the hardhat is different. (+1 for HH :) ).

If you want to double check, here is a demonstration of the issue (run foundry test -vv):

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.18;

import "forge-std/Test.sol";
import "forge-std/console.sol";

contract FooBar {
   
    function debugBytesConversion() public {
        address dummyCanonicalTokenAddress = address(this);
        uint256 chainId = 1;
        uint256 chainId2 = 2;

        bytes memory concatenation =
                bytes.concat(
                    bytes32(chainId)
        );

        console.log("Output with chain ID 1 if not keccak hashed:");
        console.log(string(abi.encodePacked(concatenation)));

        bytes memory concatenation2 =
                bytes.concat(
                    bytes32(chainId2)
                );

        console.log("Output with chain ID 2 if not keccak hashed:");
        console.log(string(abi.encodePacked(concatenation2)));


        console.log("Output with chain ID 1 if keccak hashed:");
        console.log(string(abi.encodePacked(keccak256(concatenation))));


        console.log("Output with chain ID 2 if keccak hashed:");
        console.log(string(abi.encodePacked(keccak256(concatenation2))));
    }
}

contract EmptyBytesTest is Test {
    FooBar foobar;

    function setUp() public {
        foobar = new FooBar();
    }

    function testBytes() external {
        foobar.debugBytesConversion();
    } 
}
```

Reply:
@dantaik 
The difference is coming from how `String.toString(chainId)` works under the hood, and how `bytes32(chainId)` does that.
The former with chainID = 1 will be equal to (hex"0000000000000000000000000000000000000000000000000000000000000061"), the one digit character.
The latter with chainID = 1 will be equal to (hex"0000000000000000000000000000000000000000000000000000000000000001"), the SOH (start of heading) character.

Both could work, but on the other hand we would need to adjust every call of the AddressManager's
```
 function setAddress(
        string memory name,
        address newAddress
  )
```
that using the same mechanism to construct the key (name).

Reply:
Went deeper debugging and this is how this line:
`key = string(bytes.concat(bytes32(chainId), bytes(name)));`
 interprets it:
 
<img width="666" alt="kép" src="https://user-images.githubusercontent.com/51912515/225112360-1445d853-30fd-4cd9-ad87-559c81131a7e.png">

1. Our chainId (uint256) - passed to the function- is converted to from integer to 32bytes hex number (identical number as decimal):
`0x0000000000000000000000000000000000000000000000000000000000000538`
2. The concatenation is - because the word 'taiko' is represented in ascii nibles.
`0x00000000000000000000000000000000000000000000000000000000000005387461696b6f` --> 74 61 69 6b 6f
3. After string casting it looks (and returns) like this:
`8taiko` but actually it is (with leading NULL characters, this is why editors/IDEs does not visualize them):
<img width="1118" alt="kép" src="https://user-images.githubusercontent.com/51912515/225113526-3804cf60-edfe-4a83-a664-26378a9ce1c2.png">

So this is why I would not recommend this exact solution. I'll open a branch + PR to propose a solution soon.

Reply:
Thank you for digging into it. I know the problem, I guess we just need to implement a TypeScript method for `string(bytes.concat(bytes32(chainId), bytes(name)));` then use that method in all tests.

Reply:
For example this is what we have now:
`0x00000000000000000000000000000000000000000000000000000000000005387461696b6f` --> 05 38 74 61 69 6b 6f.
05 will chopped by the string() casting within the contract and the key will be `8taiko`.

Now, we have another chan let’s say these are the created bytes:

0x00000000000000000000000000000000000000000000000000000000000002387461696b6f --> 02 38 74 61 69 6b 6f
This will result in tha same, `8taiko` .

i know that on different chains, most we will not see clashes (in case tho we want to keep track other , ‘B’ chain data on ‘A’ chain we can have) but also the public function keyForName() is queriable and not returns the proper chainId+name concatenation.

If you still want to proceed anyways, i can try to implement on test side to pass. (Today a bit sick tho.)

Reply:
I think it is dangerous especially in the bridge, where eg. in `LibBridgeProcess.sol`:
```
        address srcBridge = resolver.resolve(
            message.srcChainId,
            "bridge",
            false
        );
```

On the destination chain, checking the source chain. And in case for example destination chainId is: 1336 (0x538) and destination is : 568 (0x238). Then, this is a clash.